### DISCUSSION TITLE
PyQt Conflict

### DISCUSSION THREAD

**Discussion by randdannenberg** [mjhoptics/ray-optics#33] [Q&A] [answered]
Hi mjh and all --

I would really love to use this code. When I install it (pip install rayoptics) in the Anaconda3 iPython interface with Spyder (4.1.5), when relaunching Spyder, Spyder will not launch, and gives an error output pertaining to a conflict between two versions of PyQt and "which one to use"?

I had to delete all of Anaconda3 and reinstall all my libraries to just get back to where I was.

I am hoping there is an easy fix to the issue. I am running macOS Catalina, fyi.

Thank you very much!

**Comment by mjhoptics** [comment]
I am using a later version of spyder for my dev. Note that this isn't the latest; I couldn't get 4.2.3 to install without conflicts. Since you're using anaconda, you could install rayoptics from conda-forge rather than pip; don't think it will solve this problem but may lead to fewer conflicts. Sorry you're having problems, I've found spyder and pyqt to be very touchy.

* Spyder version: 4.2.1 None
* Python version: 3.8.6 64-bit
* Qt version: 5.12.9
* PyQt5 version: 5.12.3
* Operating System: Darwin 19.6.0

**Comment by randdannenberg** [comment]
Hi Michael,

Do I do the conda install inside Spyder-iPython?

Rand Dannenberg, Ph.D.
Owner
RaNDTek LLC
Los Angeles, CA
***@***.***
+1 727 480 9186
https://www.linkedin.com/in/randtek/
https://sites.google.com/view/randtek/home

**Reply by mjhoptics** [reply]
Hi Rand,
I have moved to using conda from the command line, rather than thru the anaconda navigator UI. A couple of reasons for this:
1) The navigator UI would often become non-responsive when I tried to update various packages
2) The default anaconda channel often had older versions of packages than were available from pip or the conda-forge channel. pyqt was one of these IIRC

I actually uninstalled anaconda and just use mini-conda now, but I think the following should work with anaconda installed also. These commands, given in a terminal window, worked for me:

`# create a new rayoptics environment named new_ro`
`conda create --name new_ro python=3.8`
`conda activate new_ro`
`conda install spyder -c conda-forge`
`conda install rayoptics -c conda-forge`
`rayoptics&`

This creates a new conda environment called new_ro, which you then activate and install spyder and rayoptics from the conda-forge channel. The rayoptics app can then be run from the command line. If you want to use jupyter lab, then install in addition:
`conda install jupyterlab -c conda-forge`
`conda install ipympl -c conda-forge`

To deactivate this environment, use:
`conda deactivate`

Here is the configuration I wound up with running the above commands.
* Spyder version: 4.2.3 None
* Python version: 3.8.8 64-bit
* Qt version: 5.12.9
* PyQt5 version: 5.12.3
* Operating System: Darwin 19.6.0

Hope this helps
Mike

**Comment by randdannenberg** [comment]
Hi – the plain old conda forge seems to have just worked.

I saw there was an online manual of sorts. Is there a repository of Python script examples of various kinds of designs? Also, what coherent capability does it have?

PS – I did the install under Spyder / iPython. All the Python examples run. However, when I type just “raytrace”, the GUI version does not just launch.

What I have seen looks great, so far!

Rand.

Rand Dannenberg Ph.D.
Owner
RaNDTek LLC
[signature_866777412]
727 480 9186<tel:727%20480%209186>
[signature_1936823298]
***@***.******@***.***>
[signature_723791032]
https://sites.google.com/view/randtek/home
[signature_763365377]
Los Angeles, CA
[Diagram, venn diagram  Description automatically generated]


From: Michael Hayford ***@***.***>
Reply-To: mjhoptics/ray-optics ***@***.***>
Date: Thursday, March 18, 2021 at 11:53 AM
To: mjhoptics/ray-optics ***@***.***>
Cc: randdannenberg ***@***.***>, Author ***@***.***>
Subject: Re: [mjhoptics/ray-optics] PyQt Conflict (#33)


Hi Rand,
I have moved to using conda from the command line, rather than thru the anaconda navigator UI. A couple of reasons for this:

  1.  The navigator UI would often become non-responsive when I tried to update various packages
  2.  The default anaconda channel often had older versions of packages than were available from pip or the conda-forge channel. pyqt was one of these IIRC

I actually uninstalled anaconda and just use mini-conda now, but I think the following should work with anaconda installed also. These commands, given in a terminal window, worked for me:

# create a new rayoptics environment named new_ro
conda create --name new_ro python=3.8
conda activate new_ro
conda install spyder -c conda-forge
conda install rayoptics -c conda-forge
rayoptics&

This creates a new conda environment called new_ro, which you then activate and install spyder and rayoptics from the conda-forge channel. The rayoptics app can then be run from the command line. If you want to use jupyter lab, then install in addition:
conda install jupyterlab -c conda-forge
conda install ipympl -c conda-forge

To deactivate this environment, use:
conda deactivate

Here is the configuration I wound up with running the above commands.

  *   Spyder version: 4.2.3 None
  *   Python version: 3.8.8 64-bit
  *   Qt version: 5.12.9
  *   PyQt5 version: 5.12.3
  *   Operating System: Darwin 19.6.0

Hope this helps
Mike

**Reply by mjhoptics** [reply]
Great, glad you hear you got a successful installation. Aside for the examples in the Read_the_Docs, I have a jupyter notebook repo (mjhoptics/ray-optics-notebooks) that has some different examples. And there is a video on YouTube that works an example [https://www.youtube.com/channel/UCbZSutJ0VrxuutyGSsylEaw].
The command to run the app is rayoptics, not raytrace. If you do this from the command line, it will include an iPython window as part of the app UI. If you run this from spyder (rayoptics/qtgui/rayopticsapp), you won't get the iPython window because of technical reasons I don't fully understand. :-}

**Comment by randdannenberg** [comment, marked as answer]
thanks very much Michael.

maybe i mistyped raytrace...

Rand Dannenberg, Ph.D.
Owner
RaNDTek LLC
Los Angeles, CA
***@***.***
+1 727 480 9186
https://www.linkedin.com/in/randtek/
https://sites.google.com/view/randtek/home

**Comment by randdannenberg** [comment]
Hi - is there a compendium of python example files of design and output cases to study?

Rand Dannenberg, Ph.D.
Owner
RaNDTek LLC
Los Angeles, CA
***@***.***
+1 727 480 9186
https://www.linkedin.com/in/randtek/
https://sites.google.com/view/randtek/home

**Comment by mjhoptics** [comment]
I have a jupyter notebook repo (mjhoptics/ray-optics-notebooks) that has some different examples. And there is a video on YouTube that works an example [https://www.youtube.com/channel/UCbZSutJ0VrxuutyGSsylEaw].

**Comment by mjhoptics** [comment]
Hi Bill,
What I wound up doing was making sure the DOE calculations were always done in air. This means for the case of the DOE on the second lens surface, I refract out of the lens and take that direction as input to the DOE. Conversely, for the DOE on the first surface, the incident ray is in air and I take the diffracted direction in air and refract it into the lens. This produces the same ray path whether I trace in the forward or backward direction. The aberration correction seems right as well; the DOE is clearly used to affect the secondary spectrum. Until I can get a validation system on production code, that’s as far as I’m taking it.

Looking forward to checking out your new UI elements!
Mike
